File: latest/week3/workshop7_m10/W07_M10_demo.py
"""W07-M10: Lab — Build a 10-Slide Data Story Deck — Python Template — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.dates import DateFormatter
os.makedirs("output", exist_ok=True)

# STEP 1: Clean theme (M06)
plt.rcParams.update({
    "axes.spines.top": False, "axes.spines.right": False,
    "axes.edgecolor": "#EEE", "axes.labelcolor": "#888",
    "xtick.color": "#888", "ytick.color": "#888",
    "grid.alpha": 0, "font.size": 10, "axes.titleweight": "bold"})

# STEP 2: Load data
nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
nf["primary_country"] = nf["country"].str.split(",").str[0].str.strip()
yearly = nf.query("2015<=year_added<=2021").groupby(["year_added","type"]).size().reset_index(name="n")
movies = yearly.query("type=='Movie'"); tv = yearly.query("type=='TV Show'")
peak_n = movies.n.max(); latest_n = movies.iloc[-1]["n"]
decline = round((1 - latest_n / peak_n) * 100)

# STEP 3: Big Idea
print("══ BIG IDEA ══")
print(f"Netflix should pivot to originals and local-language content")
print(f"because movie additions declined {decline}% while international")
print(f"genres and TV Shows drive growth.\n")

# STEP 4: Storyboard
print("══ STORYBOARD ══")
storyboard = [
    ("1", "Title", "Netflix Content Strategy", "—"),
    ("2", "Context", "8,800+ titles over a decade", "Cumulative area"),
    ("3", "Hero", f"Movie additions declined {decline}%", "Grey+accent line"),
    ("4", "Evidence 1", "International = fastest genre", "Genre grey+accent"),
    ("5", "Evidence 2", "India rose to #2", "Country bar"),
    ("6", "Evidence 3", "Seasonal release calendar", "Subseries"),
    ("7", "Mechanism", "Drivers: expansion + originals", "Ranked bar"),
    ("8", "Resolution", "Pivot budget to originals", "Action items"),
    ("9", "Variant", "[TECHNICAL] Monthly + CI", "CI ribbon"),
    ("10", "CTA", "Approve Q2 reallocation", "Action summary"),
]
for s, role, headline, chart in storyboard:
    print(f"  Slide {s}: {role:<12} | {headline:<40} | {chart}")

# STEP 5: Build slides
# Helper
def save_slide(fig, filename):
    plt.tight_layout(); fig.savefig(f"output/{filename}", dpi=300); plt.close()

# S2: Context
fig, ax = plt.subplots(figsize=(10, 5.5))
cum = nf.dropna(subset=["date_added"]).set_index("date_added").resample("ME").size().cumsum()
ax.fill_between(cum.index, cum.values, alpha=0.12, color="#1565C0")
ax.plot(cum.index, cum.values, color="#1565C0", lw=1.5)
ax.xaxis.set_major_formatter(DateFormatter("%Y"))
ax.set_title("Netflix built a catalog of 8,800+ titles over a decade", fontsize=14)
ax.set_ylabel("Cumulative Titles")
save_slide(fig, "slide02_context.png")

# S3: Hero
fig, ax = plt.subplots(figsize=(10, 5.5))
ax.plot(tv.year_added, tv.n, "-o", color="#DDD", lw=0.8, markersize=3)
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=3, markersize=7)
ax.axhline(y=peak_n, ls=":", color="#CCC", lw=0.5)
ax.axvspan(2019.5, 2021.5, alpha=0.03, color="#E53935")
ax.annotate(f"–{decline}% from peak",
    xy=(2020, (peak_n+latest_n)/2), xytext=(2017, latest_n*0.65),
    fontsize=11, fontweight="bold", color="#E53935",
    bbox=dict(boxstyle="round,pad=0.4", facecolor="white", edgecolor="#E53935", lw=1.5),
    arrowprops=dict(arrowstyle="->", connectionstyle="arc3,rad=0.3", color="#E53935", lw=1.5))
ax.text(2021.15, latest_n, " Movies", fontsize=10, fontweight="bold", color="#1565C0", va="center")
ax.text(2021.15, tv.iloc[-1]["n"], " TV Shows", fontsize=9, color="#BBB", va="center")
ax.set_xlim(2014.5, 2022.8)
ax.set_title(f"Movie additions declined {decline}% from their 2019 peak", fontsize=14)
ax.set_ylabel("Titles Added")
save_slide(fig, "slide03_hero.png")

# S4: Genre
fig, ax = plt.subplots(figsize=(10, 5.5))
genres = nf["listed_in"].str.split(",").explode().str.strip()
top5 = genres.value_counts().head(5).index
nf_g = nf.assign(listed_in=nf["listed_in"].str.split(",")).explode("listed_in")
nf_g["genre"] = nf_g["listed_in"].str.strip()
gyr = nf_g.query("genre in @top5 and 2015<=year_added<=2021").groupby(["year_added","genre"]).size().reset_index(name="n")
growth = {}
for g in top5:
    sub = gyr.query("genre==@g").sort_values("year_added")
    if len(sub)>=2 and sub.iloc[0]["n"]>0: growth[g]=sub.iloc[-1]["n"]/sub.iloc[0]["n"]
fastest = max(growth, key=growth.get) if growth else top5[0]
for g in top5:
    sub = gyr.query("genre==@g")
    c = "#2E7D32" if g==fastest else "#DDD"; lw = 2.5 if g==fastest else 0.8
    ax.plot(sub.year_added, sub.n, "-o", color=c, lw=lw, markersize=3 if c=="#DDD" else 5)
    ax.text(sub.year_added.values[-1]+0.1, sub.n.values[-1],
        f" {g[:22]}", fontsize=7, color=c if c=="#2E7D32" else "#AAA",
        fontweight="bold" if c=="#2E7D32" else "normal", va="center")
ax.set_xlim(2014.5, 2024)
ax.set_title(f"{fastest} is the fastest-growing genre category", fontsize=14)
ax.set_ylabel("Titles Added")
save_slide(fig, "slide04_genre.png")

# S5: Country
fig, ax = plt.subplots(figsize=(10, 5.5))
top10 = nf.dropna(subset=["primary_country"])["primary_country"].value_counts().head(10).sort_values()
colors = ["#E53935" if c=="India" else "#BBDEFB" for c in top10.index]
ax.barh(top10.index, top10.values, color=colors, height=0.6)
for i, (c, v) in enumerate(top10.items()):
    ax.text(v+10, i, f"{v:,}", va="center", fontsize=8, fontweight="bold")
ax.set_title("India rose to #2 in content production", fontsize=14)
save_slide(fig, "slide05_country.png")

# S6: Seasonal
fig, ax = plt.subplots(figsize=(10, 5.5))
months = np.arange(1, 13)
month_labels = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"]
for yr in range(2016, 2022):
    sub = nf.query("year_added==@yr").groupby("month_added").size().reindex(range(1,13), fill_value=0)
    c = "#E53935" if yr==2021 else "#BBB"; a = 1.0 if yr==2021 else 0.25
    ax.plot(months, sub.values, "-o", color=c, alpha=a, lw=1.5 if yr==2021 else 0.5, markersize=2 if yr<2021 else 4)
ax.set_xticks(months); ax.set_xticklabels(month_labels)
ax.set_title("Consistent seasonal calendar: Q4 holiday spike every year", fontsize=14)
ax.set_ylabel("Titles Added")
save_slide(fig, "slide06_seasonal.png")

# S7: Mechanism
fig, ax = plt.subplots(figsize=(10, 5.5))
drivers = ["COVID demand","Licensing","Local-language","Original investment","Global expansion"]
impacts = [5, 10, 22, 28, 35]
ax.barh(drivers, impacts, color="#00695C", height=0.5)
for i, v in enumerate(impacts): ax.text(v+0.5, i, f"{v}%", va="center", fontsize=9, fontweight="bold", color="#00695C")
ax.set_title("Growth driven by global expansion + original investment", fontsize=14)
ax.set_xlabel("Estimated Impact (%)")
save_slide(fig, "slide07_mechanism.png")

# S8: Resolution
fig, ax = plt.subplots(figsize=(10, 5.5)); ax.axis("off")
actions = [("1.", "20 flagship originals / quarter", "80/year"),
    ("2.", "Grow India+Korea production 50%", "+50% YoY"),
    ("3.", "Cut catalog filler 30%", "–30% YoY"),
    ("4.", "Launch 5 local-language hubs", "5 by Q4")]
for i, (num, action, target) in enumerate(actions):
    y = 0.78 - i*0.20
    ax.text(0.05, y, num, fontsize=14, fontweight="bold", color="#1565C0")
    ax.text(0.12, y, action, fontsize=12, fontweight="bold", color="#333")
    ax.text(0.12, y-0.06, f"Target: {target}", fontsize=9, color="#888", fontstyle="italic")
ax.set_title("Pivot budget: originals + local-language content", fontsize=14, pad=15)
save_slide(fig, "slide08_resolution.png")

# S9: Technical variant
fig, ax = plt.subplots(figsize=(10, 5.5))
monthly = (nf.dropna(subset=["date_added"]).query("date_added.dt.year>=2016")
    .set_index("date_added").groupby("type").resample("ME").size().reset_index(name="n"))
for t, c in [("Movie","#1565C0"),("TV Show","#E53935")]:
    sub = monthly.query("type==@t").sort_values("date_added")
    ax.scatter(sub.date_added, sub.n, s=8, c=c, alpha=0.3)
    roll = sub.n.rolling(3).mean(); roll_std = sub.n.rolling(3).std()
    ax.plot(sub.date_added, roll, color=c, lw=1.2, label=t)
    ax.fill_between(sub.date_added, (roll-1.96*roll_std).clip(lower=0), roll+1.96*roll_std, alpha=0.08, color=c)
ax.legend(fontsize=8); ax.xaxis.set_major_formatter(DateFormatter("%Y"))
ax.set_title("[TECHNICAL] Monthly Additions: 3-month rolling mean ± 1.96σ", fontsize=13)
ax.set_ylabel("Monthly Titles")
save_slide(fig, "slide09_variant.png")

print(f"\n══ DECK COMPLETE ══")
print(f"8 slide PNGs saved (slide02–slide09)")
print(f"Big Idea and storyboard printed above")
print(f"Reflection: [STUDENT WRITES 300 WORDS]")
print(f"\nAll W07-M10 Lab Python outputs saved")

File: latest/week3/workshop7_m10/W07_M10_demo.R
# ============================================================
# Workshop 7 — Module 10: Lab — Build a 10-Slide Data Story Deck
# TEMPLATE R Script — UNYT Tirana
# ============================================================
# This script is a TEMPLATE that students adapt for their dataset.
# It demonstrates the complete pipeline on Netflix.
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ═══════════════════════════════════════════════════════
# STEP 1: Set clean theme globally (M06)
# ═══════════════════════════════════════════════════════
theme_deck <- theme_minimal(base_size = 12) +
  theme(panel.grid = element_blank(),
    axis.line = element_line(color = "#EEE", linewidth = 0.3),
    plot.title = element_text(face = "bold", size = 14, lineheight = 1.2),
    plot.subtitle = element_text(color = "#666", size = 10),
    plot.caption = element_text(color = "#AAA", size = 7),
    legend.position = "none")
theme_set(theme_deck)

# ═══════════════════════════════════════════════════════
# STEP 2: Load and prepare data
# ═══════════════════════════════════════════════════════
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    month_added = month(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")))
nf_genres <- nf |> separate_rows(listed_in, sep = ", ") |>
  mutate(genre = str_trim(listed_in))

# Key metrics
yearly <- nf |>
  filter(!is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, type)
movies <- yearly |> filter(type == "Movie")
peak <- movies |> slice_max(n, n = 1)
decline <- round((1 - movies$n[movies$year_added == 2021] / peak$n) * 100)

# ═══════════════════════════════════════════════════════
# STEP 3: Big Idea (M01)
# ═══════════════════════════════════════════════════════
cat("══ BIG IDEA ══\n")
cat("Netflix should pivot investment to originals and local-language\n")
cat("content because movie additions have declined", decline, "% while\n")
cat("international genres and TV Shows drive growth.\n\n")

# ═══════════════════════════════════════════════════════
# STEP 4: Storyboard (M01 + M04 + M05)
# ═══════════════════════════════════════════════════════
cat("══ STORYBOARD ══\n")
storyboard <- tibble(
  slide = 1:10,
  role = c("Title", "Context", "Complication (Hero)", "Evidence 1",
    "Evidence 2", "Evidence 3", "Mechanism", "Resolution", "Audience Variant", "CTA"),
  headline = c(
    "Netflix Content Strategy: A Data Story",
    "Netflix built 8,800+ titles over a decade",
    paste0("Movie additions declined ", decline, "% from 2019 peak"),
    "International content is the fastest-growing genre",
    "India rose to #2 in global content production",
    "Friday releases + Q4 spikes = strategic calendar",
    "Growth driven by global expansion + original investment",
    "Pivot budget: 20 originals/quarter + 5 language hubs",
    paste0("[TECHNICAL] Monthly trend with CI ribbon (same finding)"),
    "Approve the Q2 content budget reallocation"),
  story_type = c("—", "Zoom Out", "Change", "Factors", "Zoom Out",
    "Drill Down", "Factors", "Factors", "Adaptation", "—"),
  chart = c("—", "Cumulative area", "Grey+accent line", "Genre grey+accent",
    "Country bar (India red)", "Seasonal subseries", "Driver ranked bar",
    "Action items bar", "CI ribbon (technical)", "Action summary"))
print(storyboard, n = 10)

# ═══════════════════════════════════════════════════════
# STEP 5: Build all 10 slides (M03 + M05 + M06)
# ═══════════════════════════════════════════════════════
# Slide 1: Title (text-only placeholder)
# In actual presentation: title slide with name, dataset, audience, date

# Slide 2: Context
s2 <- nf |> filter(!is.na(date_added)) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month) |> mutate(cum = cumsum(n)) |>
  ggplot(aes(x = month, y = cum)) +
  geom_area(fill = "#1565C0", alpha = 0.12) +
  geom_line(color = "#1565C0", linewidth = 1.2) +
  scale_x_date(date_breaks = "2 years", date_labels = "%Y") +
  labs(title = "Netflix built a catalog of 8,800+ titles over a decade",
       subtitle = "Cumulative titles added (2008–2021)", x = NULL, y = "Cumulative Titles")
ggsave("output/slide02_context.png", s2, width = 10, height = 5.5, dpi = 300)

# Slide 3: Hero (Complication)
s3 <- ggplot(yearly, aes(x = year_added, y = n, group = type)) +
  geom_line(color = "#DDD", linewidth = 0.8) +
  geom_line(data = movies, color = "#1565C0", linewidth = 2.5) +
  geom_point(data = movies, color = "#1565C0", size = 3) +
  geom_hline(yintercept = peak$n, linetype = "dotted", color = "#CCC") +
  annotate("rect", xmin = 2019.5, xmax = 2021.5, ymin = -Inf, ymax = Inf,
    fill = "#E53935", alpha = 0.03) +
  annotate("label", x = 2017.5, y = movies$n[movies$year_added == 2021] * 0.72,
    label = paste0("–", decline, "% from peak"),
    size = 3.5, fontface = "bold", color = "#E53935",
    fill = "white", label.size = 0.5) +
  geom_text(data = yearly |> filter(year_added == 2021),
    aes(label = type, color = type == "Movie"),
    hjust = -0.1, size = 3.5, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBB")) +
  coord_cartesian(xlim = c(2015, 2022.5)) +
  labs(title = paste0("Movie additions declined ", decline, "% from their 2019 peak"),
       subtitle = "TV Shows grew; Movies contracted — the content pipeline is shrinking",
       x = NULL, y = "Titles Added")
ggsave("output/slide03_hero.png", s3, width = 10, height = 5.5, dpi = 300)

# Slide 4: Genre evidence
top5 <- nf_genres |> count(genre, sort = TRUE) |> slice_max(n, n = 5) |> pull(genre)
genre_yr <- nf_genres |>
  filter(genre %in% top5, !is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, genre)
fastest <- genre_yr |> group_by(genre) |>
  summarise(g = last(n)/first(n), .groups = "drop") |> slice_max(g, n=1) |> pull(genre)

s4 <- ggplot(genre_yr, aes(x = year_added, y = n, group = genre)) +
  geom_line(color = "#DDD", linewidth = 0.8) +
  geom_line(data = genre_yr |> filter(genre == fastest), color = "#2E7D32", linewidth = 2.5) +
  geom_text(data = genre_yr |> group_by(genre) |> filter(year_added == max(year_added)),
    aes(label = genre, color = genre == fastest),
    hjust = -0.05, size = 2.8, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#2E7D32", "FALSE" = "#AAA")) +
  coord_cartesian(xlim = c(2015, 2023.5)) +
  labs(title = paste0(fastest, " is the fastest-growing genre category"),
       x = NULL, y = "Titles Added")
ggsave("output/slide04_genre.png", s4, width = 10, height = 5.5, dpi = 300)

# Slide 5: Country
top10 <- nf |> filter(!is.na(primary_country)) |>
  count(primary_country, sort = TRUE) |> slice_max(n, n = 10) |>
  mutate(primary_country = fct_reorder(primary_country, n))
s5 <- ggplot(top10, aes(x = primary_country, y = n,
    fill = primary_country == "India")) +
  geom_col(width = 0.6, show.legend = FALSE) +
  geom_text(aes(label = format(n, big.mark = ",")), hjust = -0.1, size = 3, fontface = "bold") +
  scale_fill_manual(values = c("TRUE" = "#E53935", "FALSE" = "#BBDEFB")) +
  coord_flip(clip = "off") +
  labs(title = "India rose to #2 in content production", x = NULL, y = "Total Titles")
ggsave("output/slide05_country.png", s5, width = 10, height = 5.5, dpi = 300)

# Slide 6: Seasonal
seasonal <- nf |> filter(!is.na(year_added), year_added >= 2016, year_added <= 2021) |>
  count(year_added, month_added) |> mutate(is_2021 = year_added == 2021)
s6 <- ggplot(seasonal, aes(x = month_added, y = n, group = year_added,
    color = is_2021, alpha = is_2021, linewidth = is_2021)) +
  geom_line() + geom_point(size = 1) +
  scale_color_manual(values = c("TRUE" = "#E53935", "FALSE" = "#BBB"), guide = "none") +
  scale_alpha_manual(values = c("TRUE" = 1, "FALSE" = 0.25), guide = "none") +
  scale_linewidth_manual(values = c("TRUE" = 1.5, "FALSE" = 0.5), guide = "none") +
  scale_x_continuous(breaks = 1:12, labels = month.abb) +
  labs(title = "Friday releases + Q4 spikes = a consistent seasonal calendar",
       subtitle = "Grey = 2016–2020, Red = 2021", x = "Month", y = "Titles")
ggsave("output/slide06_seasonal.png", s6, width = 10, height = 5.5, dpi = 300)

# Slide 7: Mechanism / Drivers
drivers <- tibble(
  driver = c("Global expansion (2016)", "Original series investment",
    "Local-language content", "Licensing partnerships", "COVID demand surge"),
  impact = c(35, 28, 22, 10, 5)) |> mutate(driver = fct_reorder(driver, impact))
s7 <- ggplot(drivers, aes(x = driver, y = impact)) +
  geom_col(fill = "#00695C", width = 0.5) +
  geom_text(aes(label = paste0(impact, "%")), hjust = -0.2, size = 3.5,
    fontface = "bold", color = "#00695C") +
  coord_flip(clip = "off") +
  labs(title = "Growth driven by global expansion + original investment",
       x = NULL, y = "Estimated Impact (%)")
ggsave("output/slide07_mechanism.png", s7, width = 10, height = 5.5, dpi = 300)

# Slide 8: Resolution
actions <- tibble(
  action = c("20 flagship originals / quarter", "Grow India+Korea 50%",
    "Cut catalog filler 30%", "Launch 5 language hubs"),
  target = c("80/yr", "+50% YoY", "–30% YoY", "5 by Q4")) |>
  mutate(action = fct_inorder(action))
s8 <- ggplot(actions, aes(x = fct_rev(action), y = 1)) +
  geom_tile(fill = "#E3F2FD", color = "white", linewidth = 2, height = 0.6) +
  geom_text(aes(label = paste0(action, "\n[", target, "]")),
    size = 3.5, fontface = "bold", lineheight = 1.2) +
  coord_flip() + theme_void(base_size = 12) +
  theme(plot.title = element_text(face = "bold", size = 14, hjust = 0.5)) +
  labs(title = "Pivot budget: originals + local-language content")
ggsave("output/slide08_resolution.png", s8, width = 10, height = 5.5, dpi = 300)

# Slide 9: Audience variant (technical version of hero)
monthly <- nf |>
  filter(!is.na(date_added), year(date_added) >= 2016) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month, type)
s9 <- ggplot(monthly, aes(x = month, y = n, color = type)) +
  geom_point(size = 0.6, alpha = 0.3) +
  geom_smooth(method = "loess", span = 0.2, se = TRUE, linewidth = 1, alpha = 0.1) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme(legend.position = "bottom",
    panel.grid.major.y = element_line(color = "#F5F5F5")) +
  labs(title = "[TECHNICAL VARIANT] Netflix Monthly Additions by Type (2016–2021)",
       subtitle = "LOESS trend ± SE ribbon | Raw monthly points | Both series shown",
       x = NULL, y = "Monthly Titles", color = NULL,
       caption = "Method: LOESS, span=0.2 | Source: netflix.csv")
ggsave("output/slide09_variant.png", s9, width = 10, height = 5.5, dpi = 300)

# Slide 10: CTA (text summary)
# In actual presentation: specific ask + contact info

# ═══════════════════════════════════════════════════════
# STEP 6: Compose deck panel
# ═══════════════════════════════════════════════════════
deck_panel <- (s2 + labs(title = "2. Context")) /
  (s3 + labs(title = "3. Hero")) /
  (s4 + labs(title = "4. Genre")) /
  (s5 + labs(title = "5. Country")) /
  (s6 + labs(title = "6. Seasonal")) /
  (s7 + labs(title = "7. Mechanism")) /
  (s8 + labs(title = "8. Resolution")) /
  (s9 + labs(title = "9. Technical Variant"))
ggsave("output/full_deck_panel.png", deck_panel, width = 10, height = 42, dpi = 150)

cat("\n══ DECK COMPLETE ══\n")
cat("8 slide PNGs saved (slide02 through slide09)\n")
cat("Composed panel: output/full_deck_panel.png\n")
cat("Storyboard: printed above\n")
cat("Big Idea: printed above\n")
cat("Reflection: [STUDENT WRITES 300 WORDS]\n")

cat("\n── All W07-M10 Lab R outputs saved ──\n")
